gym/spaces/space.py

```diff
@@ -1,7 +1,23 @@
+from typing import (
+    TypeVar,
+    Generic,
+    Optional,
+    Sequence,
+    Union,
+    Iterable,
+    Mapping,
+    Tuple,
+)
+
+import numpy as np
+
 from gym.utils import seeding
 
 
-class Space:
+T_cov = TypeVar("T_cov", covariant=True)
+
+
+class Space(Generic[T_cov]):
```

Comment:
From what I see, the only thing that the generic does is inform the signature of `sample`, correct? 

In general it'd be tempting to connect this generic with Box, Discrete and other specific spaces, which is where this type annotation would actually be useful. But I'm not sure to what extent that works in python. Maybe just defining eg. `class Box(Space[np.ndarray])`?

Reply:
Yes, `class Box(Space[np.ndarray])` is precisely how I imagined it to work. That's valid python. Making subclasses of the Space more specific, and coherent with the Env object, is the intention behind generic Env and Space:
```python
    action_space: spaces.Space[ActType]
    observation_space: spaces.Space[ObsType]
```

`gym.make` can be helped by a typing stub later on, by having overload on many literal strings we are passing. At least that can help with the most used environments. (roughly how __pow__ is handled in typeshed https://github.com/python/typeshed/pull/6287/files )